File: src/ca/joelathiessen/kaly2/main/Commander.java
package ca.joelathiessen.kaly2.main;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Commander {

  private BufferedReader br;
  private Thread robotThread;
  private Robot robot;


  public Commander(Robot robot) {
    robotThread = new Thread(robot);
    this.robot = robot;
    br = new BufferedReader(new InputStreamReader(System.in));

  }

  public void takeCommands() {
    String input;
    boolean quit = false;
    System.out.println("Enter a command (q to quit):");
    robotThread.start();
    try {

      while (quit == false) {
        input = br.readLine();
        if (input == null || input.equals("q")) {

          quit = true;
        } else if (input.equals("g")) {
          robot.startRobot();
        } else if (input.equals("r")) {
          robot.stopRobot();
        }
      }
      robot.stopRobot();

    } catch (IOException io) {
      io.printStackTrace();
    }
    robotThread.interrupt();
    
    try {
      robotThread.join();
    } catch (InterruptedException e) {
      e.printStackTrace();
      System.exit(1);
    }
    System.out.println("Program successfully completed");
  }

  public static void main(String[] args) {
    // the robot will start from here; i.e. :
    // Commander commander = new Commander(robot);
  }

}

File: src/main/kotlin/ca/joelathiessen/kaly2/Commander.kt
package ca.joelathiessen.kaly2

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class Commander(private val robot: Robot) {

    private val br: BufferedReader
    private val robotThread: Thread


    init {
        robotThread = Thread(robot)
        br = BufferedReader(InputStreamReader(System.`in`))

    }

    fun takeCommands() {
        var input: String?
        var quit = false
        println("Enter a command (q to quit, h for help):")
        robotThread.start()
        try {

            while (quit == false) {
                input = br.readLine()
                if (input == null || input == "q") {

                    quit = true
                } else if (input == "g") {
                    robot.startRobot()
                } else if (input == "r") {
                    robot.stopRobot()
                } else if (input == "h") {
                    println("Press the 'g' key to start the robot, 'r' to stop it, and 'q' to quit.")
                }
            }
            robot.stopRobot()

        } catch (io: IOException) {
            io.printStackTrace()
        }

        robotThread.interrupt()

        try {
            robotThread.join()
        } catch (e: InterruptedException) {
            e.printStackTrace()
            System.exit(1)
        }

        println("Program successfully completed")
    }

    companion object {

        @JvmStatic fun main(args: Array<String>) {
            // the robot will start itself from here; i.e. :
            // Commander commander = new Commander(robot);
            // ...

        }
    }

}
